### PR TITLE
ime: Implement virtual keyboard.

### PR DESCRIPTION
**Pull request by Zangetsu38**
- Implement SceIme with virtual keyboard.

# Result:
image

### CODE REVIEW

File: vita3k/modules/SceIme/SceIme.cpp [hunk 1]
```diff
@@ -38,7 +58,22 @@ EXPORT(int, sceImeSetText) {
...
+    if (host.ime.event_id != SCE_IME_EVENT_OPEN) {
+        SceImeEvent e{};
+        memset(&e, 0, sizeof(e));
+        e.param.rect.x = 1;
+        e.param.rect.y = 1;
+        e.param.rect.width = 1;
+        e.param.rect.height = 1;
+        e.param.caretIndex = 1;
+        e.param.text.str = Ptr<SceWChar16>(alloc(host.mem, sizeof(SceWChar16), "ime"));
+        *e.param.text.str.get(host.mem) = *host.ime.str;
```
Comment (bookmist) [inline vita3k/modules/SceIme/SceIme.cpp:70]:
```
auto str_len = strlen(host.ime.str)+1;
e.param.text.str = Ptr<SceWChar16>(alloc(host.mem, sizeof(SceWChar16)*str_len, "ime"));
strncpy(e.param.text.str.get(host.mem),host.ime.str,str_len);
```

File: vita3k/modules/SceIme/SceIme.cpp [hunk 2]
```diff
@@ -17,12 +17,32 @@
...
+#include <ime/types.h>
+
+#include <kernel/thread/thread_functions.h>
+#include <util/lock_and_find.h>
+
 EXPORT(int, sceImeClose) {
-    return UNIMPLEMENTED();
+    host.ime.ime_state = false,
+    memset(&host.ime.ime_param, 0, sizeof(SceImeParam));
+    LOG_DEBUG("sceImeClose");
```
Comment (pent0) [inline vita3k/modules/SceIme/SceIme.cpp:28]: Delete

File: vita3k/modules/SceIme/SceIme.cpp [hunk 3]
```diff
@@ -17,12 +17,32 @@
...
+    memcpy(e1.get(host.mem), e, sizeof(SceImeEvent));
+    auto thread = lock_and_find(thread_id, host.kernel.threads, host.kernel.mutex);
+    run_callback(host.kernel, *thread, thread_id, host.ime.ime_param.handler.address(), { arg.address(), e1.address() });
+    free(host.mem, e1.address());
+}
+
+EXPORT(int, sceImeOpen, SceImeParam *param) {
+    host.ime.ime_param = *param;
+    host.ime.ime_state = true;
+    LOG_DEBUG("sceImeOpen");
```
Comment (pent0) [inline vita3k/modules/SceIme/SceIme.cpp:43]: Delete

**Review by pent0 — CHANGES_REQUESTED** [review]

File: vita3k/modules/SceIme/SceIme.cpp (hunk 2 above)
Comment (Zangetsu38) [reply vita3k/modules/SceIme/SceIme.cpp:28]: Off course not worry is only for see when is called is only for debug ^^

**Review by pent0 — APPROVED** [review]

**Pull request merged**